model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteUserSessionProvider.java

```diff
@@ -462,9 +461,10 @@ private Stream<UserSessionModel> streamUserSessionByClientId(RealmModel realm, S
         var userSessionTx = getUserSessionTransaction(offline);
         return Flowable.fromIterable(QueryHelper.toCollection(userSessionIdQuery, QueryHelper.SINGLE_PROJECTION_TO_STRING))
                 .flatMapMaybe(userSessionTx::maybeGet, false, MAX_CONCURRENT_REQUESTS)
-                .concatMapMaybe(session -> maybeInitUserSessionFromQuery(session, realm, offline))
+                .blockingStream()
+                .flatMap(session -> maybeInitUserSessionFromQuery(session, realm, offline))
                 .map(UserSessionModel.class::cast)
-                .blockingStream(batchSize);
+                .limit(batchSize);
```

Comment:
This doesn't look right. The query is limited by `maxResults`; we don't need to limit the stream.

Reply:
Thanks, I think I mis-read the original code. 

Thinking about doing a query with a limit, and then removing entries that are null will lead to fewer than the maxresult entries, which will be surprising for the caller. We're doing this for entries which we retrieve and then consider them expired, as well as for those where we don't find the user. This would be a follow-up issue: #936

Reply:
I see what you mean. And, it explains why `PersistentUserSessionProvider` is fetching everything in the query. I would have to remove the limits in the Ickle query and fetch everything too :cry:  It would be a nasty piece of code :rofl: 
The link to the issue in your comment seems incorrect. 
I'll fix the implementation after my holidays.

Reply:
Link fixed. I would assume your approach is a lot more efficient - let's discuss and don't rush to an implementation. With that other issue implemented, it shouldn't happen so often.